pyomo/contrib/gdpopt/solve_master_problem.py

```diff
@@ -0,0 +1,185 @@
+                                                    tc.globallyOptimal}:
+            # we found a solution, that's all we need to keep going.
+            return tc.unbounded
+        else:
+            raise RuntimeError("Unable to find a feasible solution for the "
+                               "unbounded MILP master problem by bounding "
+                               "the objective. Either check your "
+                               "master problem initialization, or add a bound "
+                               "on the master objective value that admits a "
+                               "feasible solution.")
+
+    if terminate_cond is tc.optimal:
+        return tc.optimal
+    elif terminate_cond in {tc.locallyOptimal, tc.feasible}:
+        return tc.feasible
+    elif terminate_cond is tc.infeasible:
+        config.logger.info(
+            'MILP master problem is now infeasible. '
+            'GDPopt has finished exploring feasible discrete configurations.')
+        return tc.infeasible
+    elif terminate_cond is tc.maxTimeLimit:
+        if len(results.solution) > 0:
+            config.logger.info(
+                'Unable to optimize MILP master problem within time limit. '
+                'Using current solver feasible solution.')
+            return tc.feasible
+        else:
+            config.logger.info(
+                'Unable to optimize MILP master problem within time limit. '
+                'No solution found. Treating as infeasible, but there are no '
+                'guarantees.')
+            return tc.infeasible
+    elif (terminate_cond is tc.other and
+          results.solution.status is SolutionStatus.feasible):
+        # load the solution and suppress the warning message by setting
+        # solver status to ok.
+        config.logger.info(
+            'MIP solver reported feasible solution to MILP master problem, '
+            'but it is not guaranteed to be optimal.')
+        return tc.feasible
+    else:
+        raise ValueError(
+            'GDPopt unable to handle MILP master problem '
+            'termination condition '
+            'of %s. Solver message: %s' %
+            (terminate_cond, results.solver.message))
+
+def distinguish_mip_infeasible_or_unbounded(m, config):
+    """Distinguish between an infeasible or unbounded solution.
+
+    Linear solvers will sometimes tell me that a problem is infeasible or
+    unbounded during presolve, but not distinguish between the two cases. We
+    address this by solving again with a solver option flag on.
+
+    """
+    tmp_args = deepcopy(config.mip_solver_args)
+    if config.mip_solver == 'gurobi':
+        # This solver option is specific to Gurobi.
+        tmp_args['options'] = tmp_args.get('options', {})
+        tmp_args['options']['DualReductions'] = 0
+    mipopt = SolverFactory(config.mip_solver)
+    # gdpopt no longer supports non-auto persistent solvers, but mindtpy does,
+    # and it uses this function.
```

Comment:
Why not? Would it fail is I call a persistent solver as a MIP solver in GDPOpt?

Reply:
Yes, it will fail because it will never call `set_instance`. There has never been any real support for using a persistent solver--the previous implementation just checked if the solver was persistent, then called `set_instance` every iteration if it was. I took that out because there's no point--you can use a direct solver if that's the behavior you want.